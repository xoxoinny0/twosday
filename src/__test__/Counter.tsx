import { useCountStore, useCountAction } from "./counter.store";

export default function Counter() {
  const { count } = useCountStore();
  const { increaseCount, decreaseCount, doubleCount, halfCount } = useCountAction();
  return (
    <div style={{ width: "500px", padding: "10px", border: "1px solid green", margin: "0 auto" }}>
      <h2>Zustand Count Store</h2>
      <div data-testid="count">{count}</div>
      <button onClick={() => increaseCount()}>+</button>
      <button onClick={() => decreaseCount()}>-</button>
      <button onClick={() => doubleCount()}>x2</button>
      <button onClick={() => halfCount()}>/2</button>
    </div>
  );
}
