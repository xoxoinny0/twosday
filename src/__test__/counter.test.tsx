import { act, fireEvent, render, screen } from "@testing-library/react";
import Counter from "./counter";

describe("Counter UI", () => {
  test("카운터 페이지 스냅샷 렌더 테스트", () => {
    const counterPage = render(<Counter />);
    expect(counterPage).toMatchSnapshot();
  });

  const findTexts = [
    { id: "plusBtn", value: "+", type: "button" },
    { id: "doubleBtn", value: "x2", type: "button" },
    { id: "halfBtn", value: "/2", type: "button" },
    { id: "minusBtn", value: "-", type: "button" },
  ];

  test("카운트 값 찾고 존재하는지 확인", () => {
    render(<Counter />);

    const countEl = screen.getByTestId("count");

    expect(countEl).toBeTruthy();
  });

  test("카운터 관련 액션 버튼 찾고 존재하는지 확인", () => {
    render(<Counter />);

    findTexts.forEach(({ value }) => {
      expect(screen.getByText(value)).toBeTruthy();
    });
  });

  test("카운터 액션 버튼에 따른 카운터 값 정상반응 확인", async () => {
    render(<Counter />);

    expect(await screen.findByText(0)).toBeTruthy();

    for (const { id, value, type } of findTexts) {
      switch (id) {
        case "plusBtn":
          clickButton({ type, value });
          expect(await screen.findByText(1)).toBeTruthy();
          break;
        case "doubleBtn":
          clickButton({ type, value });
          expect(await screen.findByText(2)).toBeTruthy();
          break;
        case "halfBtn":
          clickButton({ type, value });
          expect(await screen.findByText(1)).toBeTruthy();
          break;
        case "minusBtn":
          clickButton({ type, value });
          expect(await screen.findByText(0)).toBeTruthy();
          break;
      }
    }
  });

  test("카운터 액션 버튼에 따른 카운터 값 정상반응 확인 again", async () => {
    render(<Counter />);

    expect(await screen.findByText(0)).toBeTruthy();

    for (const { id, value, type } of findTexts) {
      switch (id) {
        case "plusBtn":
          clickButton({ type, value });
          expect(await screen.findByText(1)).toBeTruthy();
          break;
        case "doubleBtn":
          clickButton({ type, value });
          expect(await screen.findByText(2)).toBeTruthy();
          break;
        case "halfBtn":
          clickButton({ type, value });
          expect(await screen.findByText(1)).toBeTruthy();
          break;
        case "minusBtn":
          clickButton({ type, value });
          expect(await screen.findByText(0)).toBeTruthy();
          break;
      }
    }
  });
});

const clickButton = async ({ type, value }: { type: string; value: string }) => {
  await act(async () => {
    fireEvent.click(await screen.findByRole(type, { name: value }));
  });
};
