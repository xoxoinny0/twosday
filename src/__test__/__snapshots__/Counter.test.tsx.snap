// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter UI 카운터 페이지 스냅샷 렌더 테스트 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="width: 500px; padding: 10px; border: 1px solid green; margin: 0px auto;"
      >
        <h2>
          Zustand Count Store
        </h2>
        <div
          data-testid="count"
        >
          0
        </div>
        <button>
          +
        </button>
        <button>
          -
        </button>
        <button>
          x2
        </button>
        <button>
          /2
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="width: 500px; padding: 10px; border: 1px solid green; margin: 0px auto;"
    >
      <h2>
        Zustand Count Store
      </h2>
      <div
        data-testid="count"
      >
        0
      </div>
      <button>
        +
      </button>
      <button>
        -
      </button>
      <button>
        x2
      </button>
      <button>
        /2
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
